File: solutions/javascript/luhn/1/luhn.js
export const valid = (number) => {
  // Remove all whitespace
  const cleaned = number.replace(/\s/g, '');
  
  // Strings of length 1 or less are not valid
  if (cleaned.length <= 1) {
    return false;
  }
  
  // Only digits are allowed after removing spaces
  if (/[^0-9]/.test(cleaned)) {
    return false;
  }
  
  let sum = 0;
  let shouldDouble = false; // start with rightmost digit not doubled
  
  // Process digits from right to left
  for (let i = cleaned.length - 1; i >= 0; i--) {
    let digit = parseInt(cleaned.charAt(i), 10);
    
    if (shouldDouble) {
      digit *= 2;
      if (digit > 9) {
        digit -= 9;
      }
    }
    
    sum += digit;
    shouldDouble = !shouldDouble;
  }
  
  return sum % 10 === 0;
};

File: solutions/go/luhn/1/luhn.go
package luhn

import (
	"strings"
	"unicode"
)

// Valid determines whether a number is valid according to the Luhn formula.
func Valid(id string) bool {
	// Strip spaces
	id = strings.Map(func(r rune) rune {
		if r == ' ' {
			return -1
		}
		return r
	}, id)

	// Length must be greater than 1 after stripping
	if len(id) <= 1 {
		return false
	}

	// All remaining characters must be digits
	for _, r := range id {
		if !unicode.IsDigit(r) {
			return false
		}
	}

	sum := 0
	double := false // start from the rightmost digit (no doubling)

	// Process digits from right to left
	for i := len(id) - 1; i >= 0; i-- {
		digit := int(id[i] - '0')
		if double {
			digit *= 2
			if digit > 9 {
				digit -= 9
			}
		}
		sum += digit
		double = !double
	}

	return sum%10 == 0
}